src/sdk/object.hpp

```diff
@@ -0,0 +1,616 @@
+                }
+            }
+
+            property_names.clear();
+            old_values.clear();
+        }
+
+        void error(std::exception_ptr err) {
+            m_block(PropertyChange<Change>(), err);
+        }
+    } callback_wrapper { std::move(callback), *this };
+
+    auto object = realm::Object(m_realm, util::demangle(typeid(Impl).name()), get_obj_key());
+    auto token = object.add_notification_callback(std::move(callback_wrapper));
+    return NotificationToken{std::move(object), std::move(token)};
+}
+
+/**
+ `Object` is a base class for model objects representing data stored in Realms.
+
+ Define your model classes by subclassing `Object` and adding properties to be managed.
+ Any managed property should have its type wrapped in the `REALM` macro. Properties
+ should then be REALM_EXPORT'd within the class body.
+
+ ```cpp
+ // dog.hpp
+ struct Dog : Object<Dog> {
+    REALM(std::string) name;
+    REALM(bool)      adopted;
+
+    REALM_EXPORT(name, adopted);
+ };
+ ```
+
+ ### Supported property types
+
+ - `std::string`
+ -  `int`, `float`, and `double`
+ - `bool`
+ - `std::chrono::time_point`
+ - `enum class`
+ - `realm::ObjectId`
+ - `X*` where  `X` is an `Object` subclass, creating a link
+ - `realm::util::Optional<X>` where `X` is a supported
+ - `EmbeddedObject`
+ - `std::vector<X*>`, where `X` is an `Object` subclass, to model many-to-many relationships.
+*/
+template <typename T>
+struct Object : ObjectBase<T>
+{
+    using base = ObjectBase<T>;
+    using self = Object<T>;
+    using derived_t = T;
+    using base::base;
+protected:
+    friend T;
+    using base::observe;
+};
+
+/**
+ MARK: Link
+ A link is an auto-updating link to another Realm Object.
+
+ Links are declared as a class member as pointers to
+ the actual Object.
+
+ class Foo;
+ class Bar : Object<Bar> {
+    REALM(Foo*) foo_link;
+ };
+ */
+template <typename T>
+struct Property<T, typename std::enable_if_t<std::is_pointer_v<T>>>
+: private PropertyBase
+{
+    using removed_pointer_t = std::remove_pointer_t<T>;
+    using base = PropertyBase;
+    using base::base;
+    using self = Object<T>;
+    using value_type = T;
+
+    ~Property()
+    {
+        m_value.reset();
+    }
+
+    Property(const Property& o)
+    {
+        *this = o;
+    }
+
+    Property& operator=(const Property& o)
+    {
+        Property::~Property();
+        if (o.m_value) {
+            new (&m_value) std::unique_ptr<removed_pointer_t>(new removed_pointer_t(*o.m_value));
+        }
+        m_unmanaged_value = o.m_unmanaged_value;
+        m_accessor = o.m_accessor;
+        return *this;
+    }
+
+    Property(Property&& o)
+    {
+        *this = std::move(o);
+    }
+
+    Property& operator=(Property&& o)
+    {
+        Property::~Property();
+        if (o.m_value) {
+            new (&m_value) std::unique_ptr<removed_pointer_t>(std::move(o.m_value));
+        }
+        m_unmanaged_value = o.m_unmanaged_value;
```

Comment:
This is super undefined behavior. Accessing the memory of an object that has been destroyed is not legal, which can happen because the placement-new above is conditional.